java/src/main/java/com/nedap/retail/example/Application.java

```diff
@@ -167,41 +173,67 @@ private static void sendHeartbeat(final BufferedReader inputBuffer) throws Excep
 
     private static void retrieveSystemInfo() throws Exception {
         final SystemInfo info = api.retrieveSystemInfo();
-        LOG.info("System info");
-        LOG.info("System id: {}", info.id);
-        LOG.info("Firmware version: {}", info.version);
-        LOG.info("System role: {}", info.systemRole);
-        LOG.info("System time: {}", info.systemTime);
+        if (info != null) {
+            LOG.info("System info");
+            LOG.info("System id: {}", info.id);
+            LOG.info("Firmware version: {}", info.version);
+            LOG.info("System role: {}", info.systemRole);
+            LOG.info("System time: {}", info.systemTime);
+        } else {
+            throw new MessageParsingException("System info could not be retrieved");
+        }
+    }
+
+    private static void retrieveGroupInfo(final GroupInfo.Group group) {
+        if (group != null) {
+            LOG.info("Group {}: '{}'", group.id, group.name);
+            if (group.units != null) {
+                for (final GroupInfo.Unit u : group.units) {
+                    LOG.info("  Unit {}: '{}'", u.id, u.name);
+                }
+            }
+            if (group.aisles != null) {
+                LOG.info("  Aisles: {}", group.aisles.stream().map((a) -> a.id).collect(Collectors.toList()));
+            }
+        }
     }
 
     private static void retrieveGroupInfo() throws Exception {
         final GroupInfo groupInfo = api.retrieveGroupInfo();
-        LOG.info("Group information");
-        for (final GroupInfo.Group group : groupInfo.groups) {
-            LOG.info("Group {}: '{}'", group.id, group.name);
-            for (final GroupInfo.Unit u : group.units) {
-                LOG.info("  Unit {}: '{}'", u.id, u.name);
+        if (groupInfo != null && groupInfo.groups != null) {
+            LOG.info("Group information");
+            for (final GroupInfo.Group group : groupInfo.groups) {
+                retrieveGroupInfo(group);
             }
-            LOG.info("  Aisles: {}", group.aisles.stream().map((a) -> a.id).collect(Collectors.toList()));
+        } else {
+            throw new MessageParsingException("Group info could not be retrieved");
```

Comment:
Is this the best place to throw the `MessageParsingException`? 
you could also throw the exception in `api.retrieveGroupInfo()`, because in that method the actual parsing is done